File: typescript/src/03-pagination.ts
/**
 * Cynco API Quickstart -- 03: Pagination
 *
 * Demonstrates three approaches to paginating through large result sets:
 * 1. Manual offset-based pagination
 * 2. Auto-pagination with the SDK's async iterator
 * 3. Cursor-based pagination for consistent results
 *
 * Run: npm run example:pagination
 */

import Cynco from "@cynco/sdk";

const cynco = new Cynco({
  apiKey: process.env.CYNCO_API_KEY,
});

/**
 * Approach 1: Manual offset-based pagination.
 * You control the offset and limit yourself.
 * Best when you need a specific page or know the total size.
 */
async function manualPagination(): Promise<void> {
  console.log("=== Manual Offset Pagination ===\n");

  const limit = 20;
  let offset = 0;
  let total = 0;
  let fetched = 0;

  do {
    const response = await cynco.invoices.list({
      limit,
      offset,
      sort: "createdAt",
      order: "desc",
    });

    total = response.pagination.total;
    fetched += response.data.length;

    console.log(
      `Page ${Math.floor(offset / limit) + 1}: ` +
        `${response.data.length} invoices (${fetched}/${total})`
    );

    offset += limit;
  } while (offset < total);

  console.log(`\nDone. Fetched ${fetched} invoices total.\n`);
}

/**
 * Approach 2: Auto-pagination with the SDK's async iterator.
 * The SDK handles fetching pages automatically -- you just iterate.
 * Best for processing all records without worrying about pagination logic.
 */
async function autoPagination(): Promise<void> {
  console.log("=== Auto Pagination (Async Iterator) ===\n");

  let count = 0;
  let totalAmount = 0;

  // The .list() method returns an object with an async iterator.
  // Use for-await-of to iterate through ALL invoices across all pages.
  for await (const invoice of cynco.invoices.list({ status: "paid" })) {
    count++;
    totalAmount += invoice.total;

    // Log every 50th invoice to show progress.
    if (count % 50 === 0) {
      console.log(`  Processed ${count} invoices so far...`);
    }
  }

  const displayTotal = (totalAmount / 100).toFixed(2);
  console.log(`\nProcessed ${count} paid invoices.`);
  console.log(`Total value: MYR ${displayTotal}\n`);
}

/**
 * Approach 3: Cursor-based pagination.
 * Uses an opaque cursor instead of offset for consistent results,
 * even if records are added or removed during pagination.
 * Best for large datasets or real-time data.
 */
async function cursorPagination(): Promise<void> {
  console.log("=== Cursor-based Pagination ===\n");

  let cursor: string | undefined = undefined;
  let pageNumber = 0;
  let totalFetched = 0;

  do {
    const response = await cynco.invoices.list({
      limit: 50,
      cursor,
    });

    pageNumber++;
    totalFetched += response.data.length;

    console.log(
      `Page ${pageNumber}: ${response.data.length} invoices ` +
        `(${totalFetched} total)`
    );

    // The nextCursor is undefined when there are no more pages.
    cursor = response.pagination.nextCursor;
  } while (cursor !== undefined);

  console.log(`\nDone. Fetched ${totalFetched} invoices total.\n`);
}

async function main(): Promise<void> {
  await manualPagination();
  await autoPagination();
  await cursorPagination();
}

main().catch((error: unknown) => {
  console.error("Pagination example failed:", error);
  process.exit(1);
});

File: python/03_pagination.py
"""
Cynco API Quickstart -- 03: Pagination

Demonstrates two approaches to paginating through large result sets:
1. Offset-based pagination (simple, good for known-size datasets)
2. Cursor-based pagination (consistent results during concurrent writes)

Run: python 03_pagination.py
"""

import os
import sys
from typing import Generator

import requests

API_KEY: str = os.environ.get("CYNCO_API_KEY", "")
BASE_URL: str = "https://app.cynco.io/api/v1"

if not API_KEY:
    print("Missing CYNCO_API_KEY environment variable.")
    print("Set it with: export CYNCO_API_KEY='cak_your_api_key_here'")
    sys.exit(1)

headers: dict[str, str] = {
    "Authorization": f"Bearer {API_KEY}",
}


def offset_pagination() -> None:
    """
    Approach 1: Offset-based pagination.
    You control the offset and limit yourself.
    """
    print("=== Offset-based Pagination ===\n")

    limit: int = 20
    offset: int = 0
    total: int = 0
    fetched: int = 0
    page: int = 0

    while True:
        response = requests.get(
            f"{BASE_URL}/invoices",
            headers=headers,
            params={
                "limit": limit,
                "offset": offset,
                "sort": "createdAt",
                "order": "desc",
            },
            timeout=30,
        )
        response.raise_for_status()

        body: dict = response.json()
        invoices: list[dict] = body["data"]
        pagination: dict = body["pagination"]

        total = pagination["total"]
        fetched += len(invoices)
        page += 1

        print(f"Page {page}: {len(invoices)} invoices ({fetched}/{total})")

        # Stop if there are no more pages.
        if not pagination["hasMore"]:
            break

        offset += limit

    print(f"\nDone. Fetched {fetched} invoices total.\n")


def iterate_all_invoices(
    status: str | None = None,
) -> Generator[dict, None, None]:
    """
    Approach 2: Generator that yields all invoices using cursor-based pagination.
    Encapsulates the pagination logic so callers just iterate.
    """
    cursor: str | None = None
    limit: int = 100

    while True:
        params: dict = {"limit": limit}
        if cursor is not None:
            params["cursor"] = cursor
        if status is not None:
            params["status"] = status

        response = requests.get(
            f"{BASE_URL}/invoices",
            headers=headers,
            params=params,
            timeout=30,
        )
        response.raise_for_status()

        body: dict = response.json()
        invoices: list[dict] = body["data"]
        pagination: dict = body["pagination"]

        for invoice in invoices:
            yield invoice

        # nextCursor is null/absent when there are no more pages.
        cursor = pagination.get("nextCursor")
        if cursor is None:
            break


def cursor_pagination() -> None:
    """Use the generator to process all paid invoices."""
    print("=== Cursor-based Pagination (Generator) ===\n")

    count: int = 0
    total_amount: int = 0

    for invoice in iterate_all_invoices(status="paid"):
        count += 1
        total_amount += invoice["total"]

        # Log progress every 50 invoices.
        if count % 50 == 0:
            print(f"  Processed {count} invoices so far...")

    display_total = total_amount / 100
    print(f"\nProcessed {count} paid invoices.")
    print(f"Total value: MYR {display_total:,.2f}\n")


def main() -> None:
    offset_pagination()
    cursor_pagination()


if __name__ == "__main__":
    main()
